File: typescript/Creational/04-Prototype/Prototype/JobCardTemplateRegistry.ts
import { JobCard } from "../Product/JobCard";

export class JobCardTemplateRegistry {
  private templates = new Map<string, JobCard>();

  register(key: string, template: JobCard): void {
    this.templates.set(key, template);
  }

  cloneTemplate(key: string): JobCard {
    const template = this.templates.get(key);

    if (!template) throw new Error(`No template registered for key '${key}'.`);

    return template.clone();
  }
}

File: csharp/Creational/04-Prototype/Prototype/JobCardTemplateRegistry.cs
public class JobCardTemplateRegistry
{
    private readonly Dictionary<string, JobCard> _templates = new();

    public void Register(string key, JobCard template) =>
        _templates[key] = template;

    public JobCard CloneTemplate(string key)
    {
        if (!_templates.TryGetValue(key, out var template))
            throw new KeyNotFoundException($"No template registered for key '{key}'.");

        return template.Clone();
    }
}
